{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no change     /home2/2459810a/miniconda3/condabin/conda\n",
      "no change     /home2/2459810a/miniconda3/bin/conda\n",
      "no change     /home2/2459810a/miniconda3/bin/conda-env\n",
      "no change     /home2/2459810a/miniconda3/bin/activate\n",
      "no change     /home2/2459810a/miniconda3/bin/deactivate\n",
      "no change     /home2/2459810a/miniconda3/etc/profile.d/conda.sh\n",
      "no change     /home2/2459810a/miniconda3/etc/fish/conf.d/conda.fish\n",
      "no change     /home2/2459810a/miniconda3/shell/condabin/Conda.psm1\n",
      "no change     /home2/2459810a/miniconda3/shell/condabin/conda-hook.ps1\n",
      "no change     /home2/2459810a/miniconda3/lib/python3.10/site-packages/xontrib/conda.xsh\n",
      "no change     /home2/2459810a/miniconda3/etc/profile.d/conda.csh\n",
      "no change     /home2/2459810a/.bashrc\n",
      "no change     /home2/2459810a/.zshrc\n",
      "no change     /home2/2459810a/.config/fish/config.fish\n",
      "no change     /home2/2459810a/.xonshrc\n",
      "no change     /home2/2459810a/.tcshrc\n",
      "No action taken.\n",
      "\n",
      "CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.\n",
      "To initialize your shell, run\n",
      "\n",
      "    $ conda init <SHELL_NAME>\n",
      "\n",
      "Currently supported shells are:\n",
      "  - bash\n",
      "  - fish\n",
      "  - tcsh\n",
      "  - xonsh\n",
      "  - zsh\n",
      "  - powershell\n",
      "\n",
      "See 'conda init --help' for more information and options.\n",
      "\n",
      "IMPORTANT: You may need to close and restart your shell after running 'conda init'.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# FILEPATH: Jupyter Notebook Cell\n",
    "!conda init --all\n",
    "# !conda create --name iCLIP_hackathon python=3.10\n",
    "!conda activate iCLIP_hackathon\n",
    "# !conda deactivate\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Various genome-related files are required during the iCLIP analysis pipeline.\n",
    "The steps to generate these should be the same across all datasets analysed, if the virus is the same, so the genome directory can be copied into the new analysis folder without needing to rerun this.\n",
    "\n",
    "Sources of data used are:\n",
    "\n",
    "Human annotation: http://ftp.ensembl.org/pub/release-106/gtf/homo_sapiens/\n",
    "Homo_sapiens.GRCh38.106.gtf.gz\n",
    "\n",
    "Human fasta: http://ftp.ensembl.org/pub/release-106/fasta/homo_sapiens/dna/ Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
    "\n",
    "SINV GTF:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat: Homo_sapiens.GRCh38.dna.primary_assembly.fa: No such file or directory\n",
      "cat: SINV_fasta.fa: No such file or directory\n",
      "/bin/bash: !gzip: command not found\n"
     ]
    }
   ],
   "source": [
    "# Merge human and SINV fasta files. Viral genome will represent an additional 'chromosome' in the genome. The fasta header used here should be the chromosome name (seqname) used in the annotation.\n",
    "\n",
    "!cd /home2/2459810a/iCLIP_analysis/PURAB_hackathon_test/GENOMEDIR\n",
    "!cat Homo_sapiens.GRCh38.dna.primary_assembly.fa SINV_fasta.fa | \\\n",
    "!gzip > HS.GRCh38.SINV.fa.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "!cd /home2/2459810a/PURAB_hackathon_test/\n",
    "!mkdir -p Data/Merged\n",
    "\n",
    "!cat \\\n",
    "/home5/castello_group/iCLIP_analysis/PURAB_analysis/Data/iClip2_Mock_200221_S0_L001_R1_001.fastq.gz \\\n",
    "/home5/castello_group/iCLIP_analysis/PURAB_analysis/Data/iClip2_Mock_200221_S0_L002_R1_001.fastq.gz \\\n",
    "/home5/castello_group/iCLIP_analysis/PURAB_analysis/Data/iClip2_Mock_200221_S0_L003_R1_001.fastq.gz \\\n",
    "/home5/castello_group/iCLIP_analysis/PURAB_analysis/Data/iClip2_Mock_200221_S0_L004_R1_001.fastq.gz >\\\n",
    "Data/Merged/PURAB_iCLIP_RAW.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## Remove PhiX Reads\n",
    "\n",
    "!/software/bbmap-v38.90/bbduk.sh \\\n",
    "in=Data/Merged/PURAB_iCLIP_RAW.fastq.gz \\\n",
    "out=Data/Merged/PURAB_iCLIP_rmPhiX.fastq.gz \\\n",
    "ref=/software/bbmap-v38.90/resources/phix174_ill.ref.fa.gz \\\n",
    "k=31 hdist=1 -Xmx16g threads=2 qin=33 \\\n",
    "stats=Data/Merged/PURAB_iCLIP_rmPhiX.stats.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## QC Sequencing Runs\n",
    "\n",
    "### Run FastQC\n",
    "\n",
    "# This is for all of the reads combined \n",
    "# The parts to focus on are: \n",
    "# *per base sequence quality - this can sometimes drop towards the 3' end. In this case, 3' ends of sequences can be trimmed. \n",
    "# *per base sequence content - biases in positions 6 to 12 will be determined by the barcodes positions 1-5 and 13-16 will reflect UMI composition\n",
    "\n",
    "\n",
    "!mkdir -p Data/Merged/FastQC\n",
    "\n",
    "files=!ls Data/Merged/*_rmPhiX.fastq.gz\n",
    "\n",
    "for file in files:\n",
    "    print(\"Processing file\", file)\n",
    "    !fastqc $file -o Data/Merged/FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: unsuccessful initial attempt using frozen solve. Retrying with flexible solve.\n",
      "Solving environment: unsuccessful attempt using repodata from current_repodata.json, retrying with next repodata source.\n",
      "\n",
      "ResolvePackageNotFound: \n",
      "  - conda==23.3.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Conda installation requirements \n",
    "# conda install -c bioconda fastqc\n",
    "!conda install -c bioconda multiqc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: multiqc: command not found\n"
     ]
    }
   ],
   "source": [
    "### Produce MultiQC report (FastQC)\n",
    "\n",
    "\n",
    "\n",
    "!mkdir -p Plots/QC\n",
    "\n",
    "!multiqc -f \\\n",
    "Data/Merged/FastQC/*_fastqc.zip \\\n",
    "-o Plots/QC \\\n",
    "-n 01_Sequencing-runs_MultiQC-report\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "#### quality filter on barcode region??? why not have this???\n",
    "\n",
    "## Demultiplex\n",
    "\n",
    "Check the frequency of barcodes\n",
    "\n",
    "This returns frequency of all barcodes in the dataset It is important to\n",
    "check that all barcodes are the top hits and also that there aren't any\n",
    "additional, over-represented barcodes In an uncontaminated library,\n",
    "expected barcodes should outnumber other barcodes by at least 10-fold\n",
    "The most frequently occurring unexpected barcodes should only differ\n",
    "slightly from barcode seqs - these are probably a result of\n",
    "amplification/ sequencing errors in the barcode region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "!zcat Data/Merged/PURAB_iCLIP_rmPhiX.fastq.gz | awk -v umi1_len=5 -v exp_bc_len=6 '{if (FNR%4==2) print substr($1,(umi1_len+1),exp_bc_len)}' | sort | uniq -c | sort -k1,1rn>barcode_freqs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "Barcode file should be a txt file with 2 columns: sample_name barcode\n",
    "\n",
    "Sample names can be whatever you want but it will make downstream steps\n",
    "simpler if it contains: - Gene name (GFP/gene of interest) - Treatment\n",
    "condition (mock/SINV) - Sample type (SMI/IP) - Replicate n\n",
    "\n",
    "e.g. DDX1_SINV_IP_rep1\n",
    "\n",
    "Barcodes should be in the format NNNNNCTGATCNNNN, with 5 Ns before the\n",
    "barcode and 4 after. The barcode sequence should be the reverse\n",
    "complement of the barcode!!! The Ns are because of the structure of\n",
    "iCLIP barcodes, which contain UMI sequences that the demultiplexing tool\n",
    "accounts for."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "!mkdir -p Trimming/Demultiplex_0MM\n",
    "\n",
    "!java -jar /software/je_1.2/je_1.2_bundle.jar demultiplex \\\n",
    "F1=Data/Merged/PURAB_iCLIP_rmPhiX.fastq.gz \\\n",
    "BF=PURAB_barcodes_rc_umi.txt \\\n",
    "RCHAR=':' \\\n",
    "O=Trimming/Demultiplex \\\n",
    "UF1=Unassigned_PURAB_iCLIP_merged_jemultiplexer.fastq \\\n",
    "M=PURAB_iCLIP_merged_jemultiplexer_out_stats.txt \\\n",
    "FASTQ_FILE_EXTENSION=fastq \\\n",
    "MM=0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## QC Individual Libraries\n",
    "\n",
    "### Run FastQC\n",
    "\n",
    "!mkdir -p Trimming/Demultiplex/FastQC\n",
    "\n",
    "for file in Trimming/Demultiplex/*.fastq.gz\n",
    "do\n",
    "echo \"Processing file $file\"\n",
    "fastqc $file -o Trimming/Demultiplex/FastQC\n",
    "echo \"\"\n",
    "done "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "# ADAPTER TRIMMING\n",
    "\n",
    "## Trim Adapters With Cutadapt\n",
    "\n",
    "Remove the adapter sequence on all of the reads and remove any sequences that are too short.\n",
    "\n",
    "NOTE: Barcode unassigned files are skipped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "mkdir -p Trimming/Adapter\n",
    "\n",
    "for file in Trimming/Demultiplex/*.fastq.gz\n",
    "do\n",
    "[[ $file == *Unassigned*.fastq.gz ]] && continue \n",
    "echo \"\" \n",
    "echo \"Processing file $file\" \n",
    "base=$(basename \"$file\") \n",
    "outfile=\"$(echo ${base} |sed -e 's/.fastq.gz/_trimmed.fastq.gz/')\" \n",
    "outshort=\"$(echo ${base} |sed -e 's/.fastq.gz/_trimmed.fastq.tooshort.gz/')\" \n",
    "outinfo=\"$(echo ${base} |sed -e 's/.fastq.gz/_trimmed.info/')\" \n",
    "cutadapt $file \\\n",
    "-a AGATCGGAAGAGCGGTTCAG \\\n",
    "-j 4 -e 0.1 -O 1 --nextseq-trim 10 --minimum-length 15 \\\n",
    "-o Trimming/Adapter/$outfile \\\n",
    "--too-short-output Trimming/Adapter/$outshort > \\\n",
    "Trimming/Adapter/$outinfo \n",
    "echo \"\" \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "## Produce MultiQC report (Cutadapt)\n",
    "\n",
    "\n",
    "mkdir -p Plots/QC \n",
    "\n",
    "multiqc -f \\\n",
    "Trimming/Adapter/*_trimmed.info \\\n",
    "-o Plots/QC \\\n",
    "-n 03_Cutadapt_MultiQC-report\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "## QC rRNA Contamination\n",
    "\n",
    "### Map to human rDNA with bbduk\n",
    "\n",
    "One thing that can reduce frequency of aligned reads is rRNA\n",
    "contamination so it's good to check what proportion of reads map to\n",
    "ribosomal RNA. Make sure that the rDNA fasta file is present in the\n",
    "analysis directory. This fasta can be downloaded from:\n",
    "<https://www.ncbi.nlm.nih.gov/nuccore/U13369.1?report=fasta>\n",
    "\n",
    "rRNA reads can also be a reflection of true binding but they will be removed during alignment as they do not uniquely map. It is useful to have an idea of mapping, therefore, as this information will be lost otherwise.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (1251040940.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[29], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    mkdir -p Trimming/rRNA\u001b[0m\n\u001b[0m             ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "mkdir -p Trimming/rRNA\n",
    "\n",
    "for file in Trimming/Adapter/*.fastq.gz\n",
    "do\n",
    "echo \"\"\n",
    "base=$(basename $file .fastq.gz)\n",
    "echo \"Processing $file\"\n",
    "/software/bbmap-v38.90/bbduk.sh \\\n",
    "-Xmx24G \\\n",
    "in=$file \\\n",
    "out=Trimming/rRNA/$(echo ${base}).minus.rRNA.fastq.gz \\\n",
    "ref=Hsap_rDNA_U13369.1.fa \\\n",
    "k=25 \\\n",
    "stats=Trimming/rRNA/$(echo ${base}).rRNA.stat.txt \\\n",
    "hdist=1\n",
    "echo \"\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "## Run STAR\n",
    "\n",
    "The original script had: \\### Gunzip fastq.gz for NTFS system to prevent\n",
    "FIFO file writing for file in Trimming/Adapter/\\*\\_trimmed.fastq.gz do\n",
    "base=$(basename \"$file\" ) output=\"Trimming/Adapter/\\$(echo \\${base}\n",
    "\\|sed -e 's/.gz//')\" gunzip -c -f \\$file done\n",
    "\n",
    "For me, this just unzipped all the files and printed them to the screen\n",
    "as far as I could tell...\n",
    "\n",
    "Instead, I unzipped the files. This seemed to work so not sure\n",
    "what the above script was written to fix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "for file in Trimming/Adapter/*_trimmed.fastq.gz \n",
    "do \n",
    "gunzip -f $file \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# Creating STAR object\n",
    "\n",
    "STAR --runThreadN 8 --runMode genomeGenerate \\\n",
    "--genomeFastaFiles HS.GRCh38.SINV.fa \\\n",
    "--sjdbGTFfile Homo_sapiens.GRCh38.106_SINV.gtf \\\n",
    "--genomeDir STAR \\\n",
    "--alignSJoverhangMin 8 --sjdbOverhang 100 \\\n",
    "--limitGenomeGenerateRAM 31000000000 --genomeSAindexNbases 10 --genomeSAsparseD 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "### Align with STAR\n",
    "\n",
    "```{bash}\n",
    "mkdir -p Alignment \n",
    "\n",
    "for file in Trimming/Adapter/*_trimmed.fastq \n",
    "do \n",
    "echo \"\" \n",
    "echo \"Processing file $file\" \n",
    "base=$(basename \"$file\") \n",
    "outdir=\"$(echo ${base} |sed -e 's/_trimmed.fastq//')\" \n",
    "mkdir Alignment/$outdir \n",
    "STAR --runMode alignReads --runThreadN 8 \\\n",
    "--outSAMtype BAM SortedByCoordinate \\\n",
    "--genomeDir GENOMEDIR/STAR \\\n",
    "--sjdbGTFfile GENOMEDIR/Homo_sapiens.GRCh38.106_HIV.modified.gtf \\\n",
    "--readFilesIn $file --outFileNamePrefix Alignment/$outdir/ \\\n",
    "--outReadsUnmapped Fastx \\\n",
    "--outFilterMismatchNmax 999 \\\n",
    "--outFilterMultimapNmax 1 \\\n",
    "--outFilterMismatchNoverLmax 0.04 \\\n",
    "--outSJfilterReads Unique \\\n",
    "--alignEndsType EndToEnd\n",
    "echo \"\" \n",
    "done \n",
    "\n",
    "```\n",
    "\n",
    "### Produce MultiQC report (STAR)\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "mkdir -p Plots/QC \n",
    "\n",
    "multiqc -f \\\n",
    "Alignment/*/*Log.final.out \\\n",
    "-o Plots/QC \\\n",
    "-n 05_STAR_MultiQC-report\n",
    "\n",
    "```\n",
    "\n",
    "### Index bam files and produce idxstats- pre-dedup\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "mkdir -p Alignment_idxstats\n",
    "\n",
    "for dir in $(ls -d Alignment/*/) \n",
    "do\n",
    "echo \"\" \n",
    "file=$dir/Aligned.sortedByCoord.out.bam \n",
    "base=$(basename \"$dir\") \n",
    "output=\"Alignment_idxstats/$(echo ${base}).idxstats.txt\"  \n",
    "echo \"Processing file $file\" \n",
    "samtools index $file \n",
    "samtools idxstats $file > $output \n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "### Produce MultiQC report (idxstat)\n",
    "\n",
    "This is pre-deduplication\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "mkdir -p Plots/QC\n",
    "\n",
    "multiqc -f \\\n",
    "Alignment_idxstats/*idxstats.txt \\\n",
    "-o Plots/QC \\\n",
    "-n 06_Chromosome-mapping-pre-dedup_MultiQC-report\n",
    "\n",
    "```\n",
    "\n",
    "# COLLAPSE UMI\n",
    "\n",
    "## Deduplicate using UMIs (je markdupes)\n",
    "\n",
    "```{bash}\n",
    "mkdir -p Dedup \n",
    "\n",
    "for dir in $(ls -d Alignment/*/) \n",
    "do\n",
    "echo \"\" \n",
    "input=$dir/Aligned.sortedByCoord.out.bam \n",
    "base=$(basename \"$dir\") \n",
    "output=\"Dedup/$(echo ${base}).dedup.je.bam\" \n",
    "outlog=\"Dedup/$(echo ${base}).dedup.je.log\" \n",
    "metrics=\"Dedup/$(echo ${base}).dedup.je.metrics\" \n",
    "java -jar /software/je_1.2/je_1.2_bundle.jar markdupes \\\n",
    "I=$input \\\n",
    "O=$output \\\n",
    "M=$metrics \\\n",
    "REMOVE_DUPLICATES=True \\\n",
    "MM=1 >\\\n",
    "$outlog\n",
    "echo \"\" \n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "### Index bam files and produce idxstats - post-dedup\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "mkdir -p Dedup_idxstats\n",
    "\n",
    "for file in $(ls Dedup/*.dedup.je.bam) \n",
    "do\n",
    "echo \"\" \n",
    "echo \"Processing file $file\" \n",
    "base=$(basename \"$file\") \n",
    "output=\"Dedup_idxstats/$(echo ${base} |sed -e 's/.dedup.je.bam/.idxstats.txt/')\" \n",
    "samtools index \"$file\" \n",
    "samtools idxstats $file > $output\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "### Produce multiQC report (idxstat)\n",
    "\n",
    "This is post-deduplication\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "mkdir -p Plots/QC \n",
    "\n",
    "multiqc -f \\\n",
    "Dedup_idxstats/*idxstats.txt \\\n",
    "-o Plots/QC \\\n",
    "-n 07_Chromosome-mapping-post-dedup_MultiQC-report\n",
    "\n",
    "```\n",
    "\n",
    "# ASSESS READ RECOVERY\n",
    "\n",
    "## Collect Read Count Data\n",
    "\n",
    "### Raw read counts\n",
    "\n",
    "the file is in the gz format so has to be unzipped before reading. zcat\n",
    "unzips the file within the operation but not outside of it (the file\n",
    "remains stored as .gz) wc -l gives the line count this is divided by 4\n",
    "since there are 4 lines per entry \\| bc is 'basic calculator' so to do\n",
    "1+2 in linux, you do 1+2 \\| bc\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "outfile='Trimming/Demultiplex/Raw-read-counts.txt' \n",
    "\n",
    "for file in $(ls Trimming/Demultiplex/*.fastq.gz)\n",
    "do\n",
    "echo \"File is $file\"\n",
    "base=$(basename \"$file\" .fastq.gz) \n",
    "count=$(echo $(zcat $file | wc -l) / 4 | bc) \n",
    "echo \"Count is $count\"\n",
    "echo \"$base $count\">> $outfile\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "### Trimmed read counts\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "outfile='Trimming/Adapter/Trimmed-read-counts.txt' \n",
    "\n",
    "for file in $(ls Trimming/Adapter/*.fastq)\n",
    "do\n",
    "echo \"File is $file\"\n",
    "chmod 744 $file\n",
    "base=$(basename \"$file\" _trimmed.fastq) \n",
    "count=$(echo $($file | wc -l) / 4 | bc) \n",
    "echo \"Count is $count\"\n",
    "echo \"$base $count\">> $outfile\n",
    "done\n",
    "\n",
    "outfile='Trimming/Adapter/Trimmed-read-counts.txt' \n",
    "\n",
    "for file in $(ls Trimming/Adapter/*.fastq)\n",
    "do\n",
    "echo \"File is $file\"\n",
    "base=$(basename \"$file\" _trimmed.fastq) \n",
    "count=$(echo $(expr $(cat $file | wc -l) / 4)) \n",
    "echo \"Count is $count\"\n",
    "echo \"$base $count\">> $outfile\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "### Uniquely mapped read counts\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "outfile='Alignment/Unique-map-read-counts.txt'\n",
    "\n",
    "for dir in $(ls -d Alignment/*/) \n",
    "do\n",
    "echo \"\" \n",
    "file=$dir/Aligned.sortedByCoord.out.bam \n",
    "base=$(basename \"$dir\") \n",
    "echo \"File is $file\" \n",
    "count=$(samtools view -c $file) \n",
    "echo \"Count is $count\"\n",
    "echo \"$base $count\" >> $outfile\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "### Deduplicated read counts\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "outfile='Dedup/Dedup-read-counts.txt'\n",
    "\n",
    "for file in $(ls Dedup/*.dedup.je.bam) \n",
    "do\n",
    "echo \"\" \n",
    "echo \"File is $file\" \n",
    "base=$(basename \"$file\" .dedup.je.bam) \n",
    "count=$(samtools view -c $file) \n",
    "echo \"Count is $count\"\n",
    "echo \"$base $count\" >> $outfile\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "# CROSSLINKED SITES\n",
    "\n",
    "These steps generate bigwig files, which are useful for determining peak shape and visualising the data. They don't, however, contribute to the core pipeline.\n",
    "\n",
    "## Get Xlsite\n",
    "\n",
    "### Collapse reads to the 5' position (I don't think this does this...)\n",
    "The 'collapsed' read is just the bam converted to bed \n",
    "Shift just shifts everything back by one position \n",
    "  The genome size file, which contains info on the length of every chromosome \n",
    "\n",
    "\n",
    "```{bash, engine.opts = '-l'}\n",
    "\n",
    "mkdir -p Xlsite/collapsed\n",
    "mkdir -p Xlsite/shifted/bed\n",
    "\n",
    "for file in Dedup/*.dedup.je.bam \n",
    "do\n",
    "echo \"Processing $file\" \n",
    "base=$(basename $file) \n",
    "outbed=\"Xlsite/collapsed/$(echo ${base} |sed -e 's/.dedup.je.bam/.collapsed.bed/')\" \n",
    "outbedshifted=\"Xlsite/shifted/bed/$(echo ${base} |sed -e 's/.dedup.je.bam/.shifted.bed/')\" \n",
    "bedtools bamtobed -i $file > $outbed \n",
    "bedtools shift -m 1 -p -1 -i $outbed -g GENOMEDIR/sizes.genome > $outbedshifted \n",
    "echo \"\"\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "## Create Bedgraphs (RPM) For Browsing\n",
    "i assume we do not require shifted for our HIV data (rozeena)\n",
    "\n",
    "```{bash, engine.opts = \"-l\"}\n",
    "mkdir -p Xlsite/shifted/bedgraph\n",
    "\n",
    "for file in Xlsite/shifted/bed/*.shifted.bed \n",
    "do \n",
    "echo \"Processing $file\" \n",
    "base=$(basename $file .shifted.bed) \n",
    "outbgplus=\"Xlsite/shifted/bedgraph/$(echo ${base}).XL.RPM.+.bedgraph\" \n",
    "outbgminus=\"Xlsite/shifted/bedgraph/$(echo ${base}).XL.RPM.-.bedgraph\" \n",
    "TmpScale=$(bc <<< \"scale=6;1000000/$(awk -v var=\"$base\" '$1 == var { print $2}' Dedup/Dedup-read-counts.txt)\") \n",
    "bedtools genomecov -bg -strand + -5 -i $file -g GENOMEDIR/sizes.genome -scale $TmpScale > $outbgplus \n",
    "bedtools genomecov -bg -strand - -5 -i $file -g GENOMEDIR/sizes.genome -scale $TmpScale > $outbgminus \n",
    "echo \"\"\n",
    "done\n",
    "```\n",
    "\n",
    "## Convert Bedgraph To BigWig\n",
    "\n",
    "### Sort bedgraphs\n",
    "\n",
    "```{bash, engine.opts = \"-l\"}\n",
    "mkdir -p Xlsite/shifted/bedgraph_sorted\n",
    "\n",
    "LC_COLLATE=C\n",
    "\n",
    "for file in Xlsite/shifted/bedgraph/*.bedgraph \n",
    "do \n",
    "echo \"Processing $file\" \n",
    "base=$(basename $file .bedgraph) \n",
    "output=\"Xlsite/shifted/bedgraph_sorted/$(echo ${base}).sorted.bedgraph\" \n",
    "sort -k1,1 -k2,2n $file > $output\n",
    "echo \"\"\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "### Convert to Bigwig\n",
    "\n",
    "```{bash, engine.opts = \"-l\"}\n",
    "mkdir -p Xlsite/shifted/bw\n",
    "\n",
    "for file in Xlsite/shifted/bedgraph_sorted/*.bedgraph \n",
    "do \n",
    "echo \"Processing $file\" \n",
    "base=$(basename $file .sorted.bedgraph) \n",
    "output=\"Xlsite/shifted/bw/$(echo ${base}).bw\" \n",
    "/software/kentUtils-v302.1.0/bin/linux.x86_64/bedGraphToBigWig –switches $file GENOMEDIR/sizes.genome $output\n",
    "echo \"\"\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "\\newpage\n",
    "\n",
    "# HTSEQ-CLIP\n",
    "\n",
    "## Extract Crosslinks\n",
    "\n",
    "Crosslink sites are extracted one position upstream of the read end as this is, theoretically, where the RT will have stopped.\n",
    "\n",
    "```{bash}\n",
    "\n",
    "mkdir -p DEW-seq/Extract\n",
    "\n",
    "for file in Dedup/*.dedup.je.bam\n",
    "do\n",
    "echo \"File is $file\"\n",
    "base=$(basename \"$file\")\n",
    "output=\"DEW-seq/Extract/$(echo ${base} |sed -e 's/.dedup.je.bam/.extract.bed/')\"\n",
    "/home4/2711498i/miniconda3/envs/htseq-clip/bin/htseq-clip extract \\\n",
    "-i $file \\\n",
    "-e 1 \\\n",
    "-s s \\\n",
    "-g -1 \\\n",
    "--ignore \\\n",
    "-o $output\n",
    "echo \"\"\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "## Creating flattened genome annotation (trial by Rozeena)\n",
    "## dont run this code snippet\n",
    "\n",
    "```{bash}\n",
    "\n",
    "gff_file=\"GENOMEDIR/Homo_sapiens.GRCh38.106_HIV.gtf\"\n",
    "echo \"File is $gff_file\"\n",
    "\n",
    "output=\"GENOMEDIR/Homo_sapiens.GRCh38.106_HIV.gtf.bed\"\n",
    "/home4/2711498i/miniconda3/envs/htseq-clip/bin/htseq-clip annotation \\\n",
    "-g $gff_file \\\n",
    "-t gene_biotype \\\n",
    "--splitExons \\\n",
    "-o $output\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Count Against Sliding Windows\n",
    "\n",
    "This requires the flattened genome annotation with the desired window size. If window size needs to be altered, this has to be done at the genome annotation level.\n",
    "\n",
    "```{bash}\n",
    "# 50/20\n",
    "mkdir -p DEW-seq/Count_w50s20\n",
    "\n",
    "for file in DEW-seq/Extract/*.bed\n",
    "do\n",
    "echo \"File is $file\"\n",
    "base=$(basename \"$file\")\n",
    "output=\"DEW-seq/Count_w50s20/$(echo ${base} |sed -e 's/.extract.bed/.count.bed/')\"\n",
    "/home4/2711498i/miniconda3/envs/htseq-clip/bin/htseq-clip count \\\n",
    "-i $file \\\n",
    "-a GENOMEDIR/sliding_window/Homo_sapiens.GRCh38.106_HIV.modified.flattened.w50s20.txt.gz \\\n",
    "-o $output\n",
    "echo \"\"\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "## Create Matrix\n",
    "\n",
    "Combine all htseq results into a single file\n",
    "\n",
    "```{bash}\n",
    "mkdir -p DEW-seq/Matrix_w50s20\n",
    "\n",
    "/home4/2711498i/miniconda3/envs/htseq-clip/bin/htseq-clip createMatrix \\\n",
    "-i DEW-seq/Count_w50s20 \\\n",
    "-e .bed \\\n",
    "-o DEW-seq/Matrix_w50s20/Full_iCLIP.matrix.txt.gz\n",
    "\n",
    "```\n",
    "\n",
    "# DEW-SEQ\n",
    "\n",
    "## R Environment\n",
    "\n",
    "From this point onwards, the analysis is done using DEWseq in R.\n",
    "\n",
    "```{r}\n",
    "library(tidyverse)\n",
    "library(data.table)\n",
    "library(IHW)\n",
    "library(DEWSeq)\n",
    "library(ggrepel)\n",
    "library(BiocParallel)\n",
    "library(scales)\n",
    "\n",
    "```\n",
    "\n",
    "```{r}\n",
    "# * * * * * Data\n",
    "\n",
    "## Read data\n",
    "countData <- fread(\"/home2/2459810a/PURAB_Analysis/DEW-seq/Matrix_w50s20/Full_iCLIP.matrix.txt.gz\", sep = \"\\t\")\n",
    "annotationData <- fread(\"/home2/2459810a/PURAB_Analysis/GENOMEDIR/sliding_window/Homo_sapiens.GRCh38.106_HIV.modified.flattened.w50s20.maptoid.txt.gz\", sep = \"\\t\")\n",
    "\n",
    "\n",
    "extract_regions <- function(resultWindows, padjThresh, log2FoldChangeThresh, trackName, fileName) {\n",
    "+   resultWindows <- as.data.frame(resultWindows)\n",
    "+   resultRegions <- extractRegions(windowRes  = resultWindows,\n",
    "+                                   padjCol    = \"p_adj_IHW\",\n",
    "+                                   padjThresh = padjThresh,\n",
    "+                                   log2FoldChangeThresh = log2FoldChangeThresh)+ toBED(windowRes = resultWindows,\n",
    "+         regionRes = resultRegions,\n",
    "+         padjCol    = \"p_adj_IHW\",\n",
    "+         padjThresh = padjThresh,\n",
    "+         log2FoldChangeThresh = log2FoldChangeThresh,\n",
    "+         trackName = trackName,\n",
    "+         fileName  = fileName\n",
    "+   )\n",
    "+   return(resultRegions)\n",
    "+ }\n",
    "> \n",
    "\n",
    "\n",
    "# * * * * * PURA\n",
    "\n",
    "##\n",
    "PURA <- countData %>%\n",
    "  dplyr::select(unique_id, contains(\"PURA\"), -contains(\"GFP\")) %>%\n",
    "  dplyr::select(unique_id, contains(\"IP\"), contains(\"SMI\")) %>%\n",
    "  column_to_rownames(var = \"unique_id\")\n",
    "\n",
    "PURA_colData <- tibble(name = colnames(PURA)) %>%\n",
    "  mutate(type = case_when(\n",
    "    str_detect(name, \"IP\") ~ \"IP\",\n",
    "    str_detect(name, \"SMI\") ~ \"SMI\"\n",
    "  )) %>%\n",
    "  mutate(type = fct_relevel(type, c(\"IP\", \"SMI\"))) %>%\n",
    "  column_to_rownames(var = \"name\")\n",
    "\n",
    "##\n",
    "PURA_ddw <- DESeqDataSetFromSlidingWindows(countData = PURA,\n",
    "                                            colData = PURA_colData,\n",
    "                                            annotObj = annotationData,\n",
    "                                            tidy = FALSE,\n",
    "                                            design = ~type,\n",
    "                                            start0based = TRUE)\n",
    "\n",
    "##\n",
    "keep <- rowSums(counts(PURA_ddw)) >= 5\n",
    "PURA_ddw <- PURA_ddw[keep,]\n",
    "rm(list = \"keep\")\n",
    "\n",
    "PURA_ddw <- estimateSizeFactors(PURA_ddw)\n",
    "sizeFactors(PURA_ddw)\n",
    "\n",
    "PURA_ddw_mRNAs <- PURA_ddw[ rowData(PURA_ddw)[,\"gene_type\"] == \"protein_coding\", ]\n",
    "PURA_ddw_mRNAs <- estimateSizeFactors(PURA_ddw_mRNAs)\n",
    "sizeFactors(PURA_ddw) <- sizeFactors(PURA_ddw_mRNAs)\n",
    "sizeFactors(PURA_ddw)\n",
    "\n",
    "PURA_ddw_tmp <- PURA_ddw\n",
    "PURA_ddw_tmp <- estimateDispersions(PURA_ddw_tmp, fitType = \"local\", quiet = TRUE)\n",
    "PURA_ddw_tmp <- nbinomWaldTest(PURA_ddw_tmp)\n",
    "tmp_significant_windows <- \n",
    "  results(PURA_ddw_tmp,\n",
    "          contrast = c(\"type\", \"IP\", \"SMI\"),\n",
    "          parallel = FALSE,\n",
    "          BPPARAM = 4,\n",
    "          tidy = TRUE,\n",
    "          filterFun = ihw) %>% \n",
    "  dplyr::filter(padj < 0.05) %>% \n",
    "  .[[\"row\"]]\n",
    "rm(\"PURA_ddw_tmp\")\n",
    "\n",
    "PURA_ddw_mRNAs <- PURA_ddw_mRNAs[ !rownames(PURA_ddw_mRNAs) %in% tmp_significant_windows, ]\n",
    "PURA_ddw_mRNAs <- estimateSizeFactors(PURA_ddw_mRNAs)\n",
    "sizeFactors(PURA_ddw) <- sizeFactors(PURA_ddw_mRNAs)\n",
    "sizeFactors(PURA_ddw)\n",
    "rm( list = c(\"tmp_significant_windows\", \"PURA_ddw_mRNAs\"))\n",
    "\n",
    "## \n",
    "PURA_ddw <- estimateDispersions(PURA_ddw, fitType = \"local\", quiet = TRUE)\n",
    "PURA_ddw <- nbinomWaldTest(PURA_ddw)\n",
    "\n",
    "PURA_resultWindows <- resultsDEWSeq(PURA_ddw,\n",
    "                                     contrast = c(\"type\", \"IP\", \"SMI\"),\n",
    "                                     parallel = FALSE,\n",
    "                                     BPPARAM = 12,\n",
    "                                     tidy = TRUE) %>% as_tibble\n",
    "\n",
    "PURA_resultWindows[,\"p_adj_IHW\"] <- adj_pvalues(ihw(pSlidingWindows ~ baseMean, \n",
    "                                                     data = PURA_resultWindows,\n",
    "                                                     alpha = 0.05,\n",
    "                                                     nfolds = 10))\n",
    "\n",
    "saveRDS(PURA_ddw, \"./DEW-seq/Matrix_w50s20/PURA_ddw_ct5.RDS\")\n",
    "saveRDS(PURA_resultWindows, \"./DEW-seq/Matrix_w50s20/PURA_resultWindows.RDS\")\n",
    "\n",
    "##\n",
    "PURA_resultWindows %>%\n",
    "  filter(p_adj_IHW < 0.05) %>%\n",
    "  filter(log2FoldChange > 2) %>%\n",
    "  filter(chromosome == \"NC_045512.2\") %>%\n",
    "  arrange(desc(log2FoldChange)) %>% pull(gene_name) %>% unique() %>% head(20)\n",
    "\n",
    "##\n",
    "pdf(\"./Plots/DEW-seq_Dispersion_PURA_ct5.pdf\", width = 7, height = 7)\n",
    "plotDispEsts(PURA_ddw)\n",
    "dev.off()\n",
    "\n",
    "#extract Regions\n",
    "\n",
    "extract_regions(resultWindows = PURA_resultWindows,\n",
    "\tpadjThresh = 0.01,\n",
    "\tlog2FoldChangeThresh = 2,\n",
    "\ttrackName = \"Binding Site HIV_PURA HIV 0.01\",\n",
    "\tfileName = \"./DEW-seq/Regions_w50s20/HIV_PURA_enrichedregions.log2FC_2.0_p0.01.bed\") %>%\n",
    "\tsaveRDS(\"./DEW-seq/Regions_w50s20/HIV_PURA_enrichedregions.log2FC_2.0_p0.01.RDS\")\n",
    "\n",
    "\n",
    "extract_regions(resultWindows = PURA_resultWindows,\n",
    "\tpadjThresh = 0.05,\n",
    "\tlog2FoldChangeThresh = 2,\n",
    "\ttrackName = \"Binding Site HIV_PURA HIV 0.05\",\n",
    "\tfileName = \"./DEW-seq/Regions_w50s20/HIV_PURA_enrichedregions.log2FC_2.0_p0.05.bed\") %>%\n",
    "\tsaveRDS(\"./DEW-seq/Regions_w50s20/HIV_PURA_enrichedregions.log2FC_2.0_p0.05.RDS\")\n",
    "\n",
    "\n",
    "# * * * * * PURB\n",
    "##\n",
    "PURB <- countData %>%\n",
    "  dplyr::select(unique_id, contains(\"PURB\"), -contains(\"GFP\")) %>%\n",
    "  dplyr::select(unique_id, contains(\"IP\"), contains(\"SMI\")) %>%\n",
    "  column_to_rownames(var = \"unique_id\")\n",
    "\n",
    "PURB_colData <- tibble(name = colnames(PURB)) %>%\n",
    "  mutate(type = case_when(\n",
    "    str_detect(name, \"IP\") ~ \"IP\",\n",
    "    str_detect(name, \"SMI\") ~ \"SMI\"\n",
    "  )) %>%\n",
    "  mutate(type = fct_relevel(type, c(\"IP\", \"SMI\"))) %>%\n",
    "  column_to_rownames(var = \"name\")\n",
    "\n",
    "##\n",
    "PURB_ddw <- DESeqDataSetFromSlidingWindows(countData = PURB,\n",
    "                                            colData = PURB_colData,\n",
    "                                            annotObj = annotationData,\n",
    "                                            tidy = FALSE,\n",
    "                                            design = ~type,\n",
    "                                            start0based = TRUE)\n",
    "\n",
    "##\n",
    "keep <- rowSums(counts(PURB_ddw)) >= 5\n",
    "PURB_ddw <- PURB_ddw[keep,]\n",
    "rm(list = \"keep\")\n",
    "\n",
    "PURB_ddw <- estimateSizeFactors(PURB_ddw)\n",
    "sizeFactors(PURB_ddw)\n",
    "\n",
    "PURB_ddw_mRNAs <- PURB_ddw[ rowData(PURB_ddw)[,\"gene_type\"] == \"protein_coding\", ]\n",
    "PURB_ddw_mRNAs <- estimateSizeFactors(PURB_ddw_mRNAs)\n",
    "sizeFactors(PURB_ddw) <- sizeFactors(PURB_ddw_mRNAs)\n",
    "sizeFactors(PURB_ddw)\n",
    "\n",
    "PURB_ddw_tmp <- PURB_ddw\n",
    "PURB_ddw_tmp <- estimateDispersions(PURB_ddw_tmp, fitType = \"local\", quiet = TRUE)\n",
    "PURB_ddw_tmp <- nbinomWaldTest(PURB_ddw_tmp)\n",
    "tmp_significant_windows <- \n",
    "  results(PURB_ddw_tmp,\n",
    "          contrast = c(\"type\", \"IP\", \"SMI\"),\n",
    "          parallel = FALSE,\n",
    "          BPPARAM = 4,\n",
    "          tidy = TRUE,\n",
    "          filterFun = ihw) %>% \n",
    "  dplyr::filter(padj < 0.05) %>% \n",
    "  .[[\"row\"]]\n",
    "rm(\"PURB_ddw_tmp\")\n",
    "\n",
    "PURB_ddw_mRNAs <- PURB_ddw_mRNAs[ !rownames(PURB_ddw_mRNAs) %in% tmp_significant_windows, ]\n",
    "PURB_ddw_mRNAs <- estimateSizeFactors(PURB_ddw_mRNAs)\n",
    "sizeFactors(PURB_ddw) <- sizeFactors(PURB_ddw_mRNAs)\n",
    "sizeFactors(PURB_ddw)\n",
    "rm( list = c(\"tmp_significant_windows\", \"PURB_ddw_mRNAs\"))\n",
    "\n",
    "## \n",
    "PURB_ddw <- estimateDispersions(PURB_ddw, fitType = \"local\", quiet = TRUE)\n",
    "PURB_ddw <- nbinomWaldTest(PURB_ddw)\n",
    "\n",
    "PURB_resultWindows <- resultsDEWSeq(PURB_ddw,\n",
    "                                     contrast = c(\"type\", \"IP\", \"SMI\"),\n",
    "                                     parallel = FALSE,\n",
    "                                     BPPARAM = 12,\n",
    "                                     tidy = TRUE) %>% as_tibble\n",
    "\n",
    "PURB_resultWindows[,\"p_adj_IHW\"] <- adj_pvalues(ihw(pSlidingWindows ~ baseMean, \n",
    "                                                     data = PURB_resultWindows,\n",
    "                                                     alpha = 0.05,\n",
    "                                                     nfolds = 10))\n",
    "\n",
    "saveRDS(PURB_ddw, \"./DEW-seq/Matrix_w50s20/PURB_ddw_ct5.RDS\")\n",
    "saveRDS(PURB_resultWindows, \"./DEW-seq/Matrix_w50s20/PURB_resultWindows.RDS\")\n",
    "\n",
    "##\n",
    "PURB_resultWindows %>%\n",
    "  filter(p_adj_IHW < 0.05) %>%\n",
    "  filter(log2FoldChange > 2) %>%\n",
    "  filter(chromosome == \"NC_045512.2\") %>%\n",
    "  arrange(desc(log2FoldChange)) %>% pull(gene_name) %>% unique() %>% head(20)\n",
    "\n",
    "##\n",
    "pdf(\"./Plots/DEW-seq_Dispersion_PURB_ct5.pdf\", width = 7, height = 7)\n",
    "plotDispEsts(PURB_ddw)\n",
    "dev.off()\n",
    "\n",
    "#extract Regions\n",
    "\n",
    "extract_regions(resultWindows = PURB_resultWindows,\n",
    "\tpadjThresh = 0.01,\n",
    "\tlog2FoldChangeThresh = 2,\n",
    "\ttrackName = \"Binding Site HIV_PURB HIV 0.01\",\n",
    "\tfileName = \"./DEW-seq/Regions_w50s20/HIV_PURB_enrichedregions.log2FC_2.0_p0.01.bed\") %>%\n",
    "\tsaveRDS(\"./DEW-seq/Regions_w50s20/HIV_PURB_enrichedregions.log2FC_2.0_p0.01.RDS\")\n",
    "\n",
    "\n",
    "extract_regions(resultWindows = PURB_resultWindows,\n",
    "\tpadjThresh = 0.05,\n",
    "\tlog2FoldChangeThresh = 2,\n",
    "\ttrackName = \"Binding Site HIV_PURB HIV 0.05\",\n",
    "\tfileName = \"./DEW-seq/Regions_w50s20/HIV_PURB_enrichedregions.log2FC_2.0_p0.05.bed\") %>%\n",
    "\tsaveRDS(\"./DEW-seq/Regions_w50s20/HIV_PURB_enrichedregions.log2FC_2.0_p0.05.RDS\")\n",
    "\n",
    "\n",
    "# * * * * * Mock\n",
    "\n",
    "##\n",
    "Mock <- countData %>%\n",
    "  dplyr::select(unique_id, contains(\"Mock\"), -contains(\"IgG\")) %>%\n",
    "  dplyr::select(unique_id, contains(\"IP\"), contains(\"SMI\")) %>%\n",
    "  column_to_rownames(var = \"unique_id\")\n",
    "\n",
    "Mock_colData <- tibble(name = colnames(Mock)) %>%\n",
    "  mutate(type = case_when(\n",
    "    str_detect(name, \"IP\") ~ \"IP\",\n",
    "    str_detect(name, \"SMI\") ~ \"SMI\"\n",
    "  )) %>%\n",
    "  mutate(type = fct_relevel(type, c(\"IP\", \"SMI\"))) %>%\n",
    "  column_to_rownames(var = \"name\")\n",
    "\n",
    "##\n",
    "Mock_ddw <- DESeqDataSetFromSlidingWindows(countData = Mock,\n",
    "                                            colData = Mock_colData,\n",
    "                                            annotObj = annotationData,\n",
    "                                            tidy = FALSE,\n",
    "                                            design = ~type,\n",
    "                                            start0based = TRUE)\n",
    "\n",
    "##\n",
    "keep <- rowSums(counts(Mock_ddw)) >= 5\n",
    "Mock_ddw <- Mock_ddw[keep,]\n",
    "rm(list = \"keep\")\n",
    "\n",
    "Mock_ddw <- estimateSizeFactors(Mock_ddw)\n",
    "sizeFactors(Mock_ddw)\n",
    "\n",
    "Mock_ddw_mRNAs <- Mock_ddw[ rowData(Mock_ddw)[,\"gene_type\"] == \"protein_coding\", ]\n",
    "Mock_ddw_mRNAs <- estimateSizeFactors(Mock_ddw_mRNAs)\n",
    "sizeFactors(Mock_ddw) <- sizeFactors(Mock_ddw_mRNAs)\n",
    "sizeFactors(Mock_ddw)\n",
    "\n",
    "Mock_ddw_tmp <- Mock_ddw\n",
    "Mock_ddw_tmp <- estimateDispersions(Mock_ddw_tmp, fitType = \"local\", quiet = TRUE)\n",
    "Mock_ddw_tmp <- nbinomWaldTest(Mock_ddw_tmp)\n",
    "tmp_significant_windows <- \n",
    "  results(Mock_ddw_tmp,\n",
    "          contrast = c(\"type\", \"IP\", \"SMI\"),\n",
    "          parallel = FALSE,\n",
    "          BPPARAM = 4,\n",
    "          tidy = TRUE,\n",
    "          filterFun = ihw) %>% \n",
    "  dplyr::filter(padj < 0.05) %>% \n",
    "  .[[\"row\"]]\n",
    "rm(\"Mock_ddw_tmp\")\n",
    "\n",
    "Mock_ddw_mRNAs <- Mock_ddw_mRNAs[ !rownames(Mock_ddw_mRNAs) %in% tmp_significant_windows, ]\n",
    "Mock_ddw_mRNAs <- estimateSizeFactors(Mock_ddw_mRNAs)\n",
    "sizeFactors(Mock_ddw) <- sizeFactors(Mock_ddw_mRNAs)\n",
    "sizeFactors(Mock_ddw)\n",
    "rm( list = c(\"tmp_significant_windows\", \"Mock_ddw_mRNAs\"))\n",
    "\n",
    "## \n",
    "Mock_ddw <- estimateDispersions(Mock_ddw, fitType = \"local\", quiet = TRUE)\n",
    "Mock_ddw <- nbinomWaldTest(Mock_ddw)\n",
    "\n",
    "Mock_resultWindows <- resultsDEWSeq(Mock_ddw,\n",
    "                                     contrast = c(\"type\", \"IP\", \"SMI\"),\n",
    "                                     parallel = FALSE,\n",
    "                                     BPPARAM = 12,\n",
    "                                     tidy = TRUE) %>% as_tibble\n",
    "\n",
    "Mock_resultWindows[,\"p_adj_IHW\"] <- adj_pvalues(ihw(pSlidingWindows ~ baseMean, \n",
    "                                                     data = Mock_resultWindows,\n",
    "                                                     alpha = 0.05,\n",
    "                                                     nfolds = 10))\n",
    "\n",
    "saveRDS(Mock_ddw, \"./DEW-seq/Matrix_w50s20/Mock_ddw_ct5.RDS\")\n",
    "saveRDS(Mock_resultWindows, \"./DEW-seq/Matrix_w50s20/Mock_resultWindows.RDS\")\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
