---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
library(rtracklayer)
library(data.table)
library(clusterProfiler)
library(org.Hs.eg.db)
library(openxlsx)
library(UpSetR)
library(ggrepel)
library(GenomicRanges)
library(plyranges)
library(ggseqlogo)
library(BSgenome.Hsapiens.UCSC.hg38)
library(universalmotif)
library(ComplexHeatmap)
library(patchwork)

hs <- org.Hs.eg.db

source("iCLIP_analysis_functions_2.R")
source("cliprofiler_functions.R")

load("genome_annotations.gtf")
load("genome_annotation_introns.gtf")
```

#Load files
```{r}
PURA_files <- list.files(pattern = "*_PURA_enrichedregions.log2FC_2.0_p0.01.bed")
```

#Convert to granges files and extract binding site information
```{r}

PURA_bs_gr <- map2(PURA_files, 
                         c("IP","SMI"),
                         ~ get_gr_and_bs(.x, "PURA", .y))
```

#Identify gene region binding preference
```{r}

#as counts
PURA_bs_gr %>% 
map_dfr(~ .x[["bs"]]) %>% 
  #dplyr::filter(gene_biotype == "protein_coding") %>% 
  dplyr::filter(type %in% c("five_prime_utr", "CDS", "three_prime_utr", "intron")) %>% 
  distinct(id, type, condition) %>% 
  group_by(condition) %>% 
  count(type) %>%
  #dplyr::mutate(percent = n*100/sum(n)) %>% 
  dplyr::mutate(condition = factor(condition, levels = c("IP","SMI"))) %>% 
  ggplot(aes(x = condition, y = n, fill = type)) + geom_col(position = "dodge",colour = "black") +
  theme_bw()

#as proportions
PURA_bs_gr %>% 
map_dfr(~ .x[["bs"]]) %>% 
  dplyr::filter(type %in% c("five_prime_utr", "CDS", "three_prime_utr", "intron")) %>% 
  distinct(id, type, condition) %>% 
  group_by(condition) %>% 
  count(type) %>%
  dplyr::mutate(percent = n*100/sum(n)) %>% 
  dplyr::mutate(condition = factor(condition, levels = c("IP","SMI"))) %>% 
  ggplot(aes(x = condition, y = percent, fill = type)) + geom_col(colour = "black") +
  theme_bw()
```

#Generate meta-gene profile of all binding sites

This step requires an additional genome annotation and functions from the cliProfiler package.
```{r}
anno <- rtracklayer::import.gff3(con = "/Users/louieiselin/Downloads/gencode.v41.annotation.gff3")

metagenes <- PURA_bs_gr %>% map(~ metaGeneProfile(.x[["gr"]], anno, split = FALSE))

smi_meta <- metagenes[[1]]$Plot + theme(legend.position = "none") + ggtitle("SMI")
meta4h <- metagenes[[3]]$Plot + theme(legend.position = "none")+ ggtitle("IP")


mock_meta + meta4h + meta18h
```

#Gene biotype preference
```{r}

map_dfr(PURA_bs_gr, function(x){
  
pseudo <- grep("pseudogene", x[["bs"]]$gene_biotype, value = TRUE)
  
x[["bs"]] %>% 
  dplyr::mutate(gene_biotype = ifelse(gene_biotype %in% pseudo, "pseudogene", gene_biotype)) %>% 
  distinct(id, gene_biotype, condition) %>% 
  group_by(condition) %>% 
  count(gene_biotype) %>%
  dplyr::mutate(percent = n*100/sum(n)) %>% 
  dplyr::arrange(percent) %>% 
  dplyr::mutate(gene_biotype = factor(gene_biotype, levels = gene_biotype))

}) %>% 
  dplyr::mutate(condition = factor(condition, levels = c("IP","SMI"))) %>% 
  ggplot(aes(x = gene_biotype, y = percent, fill = gene_biotype)) + 
  geom_col(colour = "black") + coord_flip() + facet_wrap(~condition) + theme_bw() +
  theme(legend.position = "none") 
```

#Information on number of binding sites and target genes
```{r}
count_info <- XRN1_bs_gr %>% map_dfr(~ get_bs_count_info(.x[["bs"]], .x[["gr"]]))

count_info %>% 
  dplyr::filter(type != "n_unique_binding_sites") %>% 
  dplyr::mutate(type = ifelse(type == "n_unique_binding_sites_nonoverlapping", "binding sites",
                              "target genes")) %>% 
  dplyr::mutate(condition = factor(condition, levels = c("Mock", "4h", "18h"))) %>% 
  ggplot(aes(x = condition, y = count, fill = type)) + 
  geom_col(position = "dodge",colour = "black") +
  theme_bw() + scale_fill_manual(values = c("#6F1B45", "#1B6F1B"))
```

#Upset plot of overlaps between different groups
```{r}
gene_lists <- XRN1_bs_gr %>% 
  map(function(x){
    
    x[["bs"]] %>% pull(gene_name) %>% unique()
    
  }) 

names(gene_lists) <- c("Mock", "18h", "4h")
  
  
upset(fromList(gene_lists), order.by = "freq")
```

#Venn diagram of overlaps between different samples
A venn diagram is better when there are fewer comparisons to make
```{r}

pdf("overlap_venn.pdf", width = 1.7, height = 3)
plot(eulerr::euler(gene_lists, shape = "ellipse"), 
     quantities = list(cex = 1),
     fills = c("#6a3d9a", "#1f78b4", "#489FA7"))
dev.off()

```

#GO enrichment analysis for target genes
```{r}
#the organism database is loaded at the top as a package would be
e_ids_XRN1_mock_bound <- AnnotationDbi::select(hs, 
       keys = gene_lists[[2]],
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL") %>% pull(ENTREZID)

#BP (biological process) can be changed to MF (molecular function) or CC (cellular compartment)
ego_XRN1_mock_bound <- enrichGO(
                gene          = e_ids_XRN1_mock_bound,
                #universe      = names(geneList),
                OrgDb         = hs,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable      = TRUE)



#many GO terms are overlapping so this function seeks to combine overlapping terms and simplify the output
#the cutoff can be adjusted, depending on how many GO terms you want/ how similar they are
ego_XRN1_mock_bound_simp <- simplify(ego_XRN1_mock_bound, cutoff=0.6, by="p.adjust", select_fun=min)

ego_XRN1_mock_bound_simp %>% as.tibble()%>% 
  tidyr::separate(GeneRatio, c("gene_go", "gene_no_go"))%>% 
  tidyr::separate(BgRatio, c("back_go", "back_no_go")) %>% 
  dplyr::mutate_at(c("gene_go", "gene_no_go", "back_go", "back_no_go"), ~ as.numeric(.)) %>% 
  dplyr::mutate(gene_no_go = gene_no_go - gene_go, back_no_go = back_no_go - back_go) %>% 
  dplyr::mutate(odds_ratio = (gene_go/gene_no_go)/(back_go/back_no_go)) %>% 
  ggplot() + geom_point(aes(x = Description, y = "1", size = log2(odds_ratio), colour = -log10(p.adjust))) + coord_flip() + theme_bw() + theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.title.y=element_blank(),
        axis.ticks.x=element_blank(),
         #plot.margin=grid::unit(c(0,0,0,0), "mm")#, 
        #legend.position = "none"
        ) + 
  scale_colour_gradient2(
    low = "#f2f0f7", 
    mid = "#9e9ac8", 
    high = "#54278f",
    midpoint = 1e-06
  )

```


#Plot coverage of signal across the viral genome
```{r}
SINV_coverage <- read.csv("XRN1_SINV_coverage.csv")

SINV_coverage_means <- SINV_coverage %>% 
  group_by(condition, X2, X3, sample) %>% dplyr::summarise(mean = mean(X4)) 

SINV_coverage_means %>% 
  dplyr::filter(sample == "SINV18h") %>% 
  pivot_wider(names_from = condition, values_from = mean) %>% 
  dplyr::mutate(SMI = ifelse(is.na(SMI), 0, SMI )) %>% 
  dplyr::mutate(normalised = IP - SMI) %>%
  dplyr::mutate(direction = ifelse(normalised < 0, "SMI", "IP")) %>% 
   ggplot(aes(x = X3, y = normalised, colour = direction)) +
  geom_line( size = 1) + scale_colour_manual(values = c("#2c7fb8", "grey80"))
```

